stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -14,434 +14,241 @@
-    if (!is_constant<T>::value)
-      return stan::math::num_elements(op_value_);
-    return 0;
-  }
-
-  inline void build(vari** varis, double* partials) const {
-    if (!is_constant<T>::value) {
-      register_operands(op_, varis);
-      std::copy(partials_.data(), partials_.data() + partials_.size(),
-                partials);
+    template <typename... Pargs>
+    void accumulate_adjoints(double* dest, const var& x, const Pargs&... args) {
+      *dest += x.adj();
+      accumulate_adjoints(dest + 1, args...);
     }
-  }
-};
-
-/* for types which are constants (non-vars), the implemtation can be
- * seriously simplified and based on constant refs being returned
- */
-
-template <>
-struct local_operand_and_partials<double> {
-  using Op_t = std::vector<double>;
-  using Op_local_t = const Op_t&;
-  using Op_local_slice_t = Op_t;
-
-  const Op_t& op_;
-
-  explicit local_operand_and_partials(const Op_t& op) : op_(op) {}
-
-  local_operand_and_partials(const local_operand_and_partials<double>& other)
-      : op_(other.op_) {}
-
-  inline void add_local_adjoint(const Op_t& local_op, std::size_t offset = 0) {}
-
-  inline void add_other_adjoint(
-      const local_operand_and_partials<double>& other) {}
-
-  inline Op_local_t local_op() const { return op_; }
-
-  inline const Op_local_slice_t local_op(std::size_t start,
-                                         std::size_t end) const {
-    const Op_local_slice_t slice(op_.begin() + start, op_.begin() + end);
-    return slice;
-  }
-
-  inline std::size_t num_elements() const { return 0; }
-
-  inline void build(vari** varis, double* partials) const {}
-};
-
-template <>
-struct local_operand_and_partials<int> {
-  using Op_t = std::vector<int>;
-  using Op_local_t = const Op_t&;
-  using Op_local_slice_t = Op_t;
-
-  const Op_t& op_;
 
-  explicit local_operand_and_partials(const Op_t& op) : op_(op) {}
-
-  local_operand_and_partials(const local_operand_and_partials<int>& other)
-      : op_(other.op_) {}
-
-  inline void add_local_adjoint(const Op_t& local_op, std::size_t offset = 0) {}
-
-  inline void add_other_adjoint(const local_operand_and_partials<int>& other) {}
-
-  inline Op_local_t local_op() const { return op_; }
-
-  inline const Op_local_slice_t local_op(std::size_t start,
-                                         std::size_t end) const {
-    const Op_local_slice_t slice(op_.begin() + start, op_.begin() + end);
-    return slice;
-  }
-
-  inline std::size_t num_elements() const { return 0; }
-
-  inline void build(vari** varis, double* partials) const {}
-};
+    // Works with anything that has operator[Integral] defined
+    template <typename... Pargs, typename Vec,
+              require_vector_like_vt<is_var, Vec>...>
+    void accumulate_adjoints(double* dest, const Vec& x, const Pargs&... args) {
+      for (size_t i = 0; i < x.size(); ++i) {
+        dest[i] += x[i].adj();
+      }
+      accumulate_adjoints(dest + x.size(), args...);
+    }
```

Comment:
So I know these look real weird in the templates but essentially this works with any object that has a `operator[int]` defined. So now we just need to write a `get_vari()` and `total_size()` function and this can traverse vectors of eigen matrices of vars and yadayada

Reply:
Explain more?

I was thinking we could redo this like:

```
template<typename R, typename... Pargs>
double* accumulate_adjoints(double *dest, const std::vector<R>& x, const Pargs&... args) {
  for(size_t i = 0; i < x.size(); ++i) {
    dest = accumulate_adjoints(dest, x[i]);
  }
  return accumulate_adjoints(dest + x.size(), args...);
}
```

And all the other accumulates return the updated dest pointer and then this'd recursively work for arrays of arrays of arrays etc.

We'd might then want an explicit:
```
template<typename... Pargs>
double* accumulate_adjoints(double *dest, const std::vector<var>& x, const Pargs&... args) {
  for(size_t i = 0; i < x.size(); ++i) {
    dest[i] += x[i].adj();
  }
  accumulate_adjoints(dest + x.size(), args...);
}
```

Cause that seems like it'd be faster, but I dunno if it'd matter? So maybe not since it's just more complexity.

Reply:
+1 ... yup. This recursive way of extracting results and moving forward the pointer is how my original idea worked. A bit involved, but I did like it.

Reply:
Ah sorry yes I think your above makes sense. If 

```cpp
dest = accumulate_adjoints(dest, x[i]);
```

gets inlined then idt it would be much of a slowdown